_posts/2024-02-09-Database-interaction-with-skunk-scala-library:-A-comprehensive-guide.md

```diff
@@ -0,0 +1,555 @@
+
+1. `Temporal`
+  * **Purpose**: Handle conversions between Scala types and PostgreSQL temporal types like timestamp, date, time, etc.
+  * **Use**: When executing queries or fetching data from the database, temporal instances ensure that the date and time values retrieved from PostgreSQL are correctly mapped to Scala types, and vice versa. For example, a PostgreSQL `timestamp` would be mapped to a Scala `java.time.Instant`
+
+2. `Trace`
+  * **Purpose**: Responsible for distributed tracing, allowing us to monitor the flow of requests across services and components.
+  * **Use**: Tracing provides insights into database interactions, helping to identify bottlenecks, failures, or performance issues within our application. You can integrate Skunk with your preferred tracing system, such as [Zipkin](https://zipkin.io/pages/quickstart) by providing appropriate Trace instances.
+
+3. `Network`
+  * **Purpose**: Handle network operations related to database connections.
+  * **Use**: Network instances manage the underlying network connections to the PostgreSQL server. They handle tasks such as establishing connections, pooling, and managing the lifecycle of connections. This ensures efficient and reliable communication between our application and the database.
+
+4. `Console`
+  * **Purpose**: Provides logging functionality to output messages to the console.
+  * **Use**: Console instances are useful for logging debug information, errors, or other messages related to database interactions. They enable us to monitor the behavior of our application and diagnose issues during development and debugging. These messages can be logged to the console or redirected to other logging systems as needed.
+
+  In summary, these implicit instances (`Temporal, Trace, Network, Console`) are crucial for configuring and managing database connections in Skunk as they ensure correct data handling.
+
+  Hopefully we uunderstand the roles of the above instances in creating a skunk database connection, let's use `IOApp` from cats-effect as a context to create a connection and test our connection by simply querying our current date and time.
+  ```scala
+      ...
+      import cats.effect.{ExitCode, IO, IOApp}
+      import skunk.*
+      import skunk.implicits.*
+      import skunk.codec.all.*
+      import natchez.Trace.Implicits.noop
+
+      object main extends IOApp {
+
+        override def run(args: List[String]): IO[ExitCode] =
+          val config: Either[ConfigReaderFailures, Config] = ConfigSource.default.load[Config]
+          config match
+            case Left(configFailure) =>
+              for {
+                _ <- IO(println(configFailure.prettyPrint(2)))
+              } yield ExitCode.Success
+
+            case Right(configValues) =>
+              val resource: Resource[IO, Session[IO]] = new DbConnection[IO].single(configValues)
+              val resources: Resource[IO, Resource[IO, Session[IO]]] = new DbConnection[IO].pooled(configValues)
```

Comment:
let's separate two methods where in each we use one version of the resource, instead of demonstrating both here.